Week2

AmirhosseinOlyaei opened the pull request.

AmirhosseinOlyaei committed:
answers for lesson 1

AmirhosseinOlyaei committed:
initial setup

AmirhosseinOlyaei committed:
writeWithPromisesAwait

AmirhosseinOlyaei committed:
Added stream processing and improved event emission

Introduced a new example demonstrating stream processing with custom configuration to efficiently read large files. Also implemented examples showcasing the use of Node.js EventEmitter, including basic event handling, event-driven async flows, and chaining of events, which fosters understanding of event-driven patterns in Node.js. Additionally, enhanced server request logging for better monitoring. Improved file writing with Promises for cleaner asynchronous code execution.

mattcarlotta commented on 01-node-tutorial/answers/customEmitter.js:
```diff
@@ -0,0 +1,43 @@
+};
+
+doWait();
+emitter.emit("eventWait", "Hello World!");
```
Requires change:
This is an odd way (ChatGPT like) to solve the homework assignment. You're currently using a Promise to return the result of the event message while also listening for the event. Instead, you can refactor it to trigger an event after a short delay, then listen for that event and log it:
```js
// this is called an IFFE: https://developer.mozilla.org/en-US/docs/Glossary/IIFE
(async () => {
  // create an artificial 2 second delay
  await new Promise(res => setTimeout(res, 2000));
  // emit a custom event
  emitter.emit("my-custom-event", "Hello World!");
})();

// listen for custom event and log the result to the console
emitter.on("my-custom-event", event => console.log("We got an event! Here it is: ", event));
```
This pattern is called a [Pub/Sub](https://cloud.google.com/pubsub/docs/overview) which allows you to emit events from anywhere in your codebase and then listen for those events in one or many places.

mattcarlotta commented on 01-node-tutorial/answers/writeWithPromisesAwait.js:
```diff
@@ -0,0 +1,26 @@
+
+async function writer() {
+  try {
+    await writeFile("temp.txt", "Line 1\nLine 2\nLine 3\n");
```
Requires changes:
I believe the assignment wanted you to write to the file 3 times, instead of using `\n`. This would resemble the `writeWithPromisesThen` example, where you write to the same file 3 times as well.
```js
await writeFile("...", "...");
await writeFile("...", "...", { flag: "a" });
await writeFile("...", "...", { flag: "a" });
```

mattcarlotta reviewed (changes requested):
Excellent work, Amir! You did extremely well on this assignment. However, before I can approve this PR, there are a few callouts below that need to be addressed.

### When you've addressed the feedback below, please do NOT resubmit your assignment using the CTD form. Instead, send me a DM when this PR is ready for another review.

AmirhosseinOlyaei committed:
Refactor event-waiting logic with timeout

Replaced the previous implementation with an immediately-invoked function expression (IIFE) to streamline event-waiting logic. This change introduces a deliberate 2-second delay before emitting a custom event to simulate asynchronous behavior in a more practical scenario. Simplified code improves readability and decouples event waiting from the rest of the codebase.

AmirhosseinOlyaei committed:
Optimize file writes with append mode

Refactored the file write operation to progressively append lines rather than writing all at once. By using the append flag, we reduce memory usage for large file writes and prevent potential data loss between writes. This change enhances the reliability and efficiency of the file output process.

mattcarlotta reviewed (approved):
Looks good to me! Excellent work, Amir! 👍
